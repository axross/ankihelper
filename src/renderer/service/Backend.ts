import { Event, IpcRenderer } from 'electron';
import { v4 } from 'uuid';
import Configuration from '../../common/model/Configuration';
import Entry from '../../common/model/Entry';
import File from '../../common/model/File';

class Backend {
  private ipcRenderer: IpcRenderer;

  public async getIpa(word: string): Promise<string> {
    const trimmed = word.trim();

    if (trimmed.length === 0) {
      return '';
    }

    const words = trimmed.split(' ');

    const ipas = await Promise.all(
      words.map(
        w =>
          new Promise((resolve, reject) => {
            const hash = v4();

            const onRespond = (_: Event, ipa: string) => {
              this.ipcRenderer.removeListener(`error#${hash}@ipa.get`, onError);

              resolve(ipa);
            };

            const onError = (_: Event, errorMessage: string) => {
              this.ipcRenderer.removeListener(`respond#${hash}@ipa.get`, onRespond);

              reject(new Error(errorMessage));
            };

            this.ipcRenderer.once(`respond#${hash}@ipa.get`, onRespond);
            this.ipcRenderer.once(`error#${hash}@ipa.get`, onError);

            this.ipcRenderer.send('ipa.get', hash, w);
          })
      )
    );

    return ipas.join(' ');
  }

  public getEntries(text: string): Promise<Entry[]> {
    if (text === '') {
      return Promise.resolve([]);
    }

    return new Promise((resolve, reject) => {
      const hash = v4();

      const onRespond = (_: Event, entries: Entry[]) => {
        this.ipcRenderer.removeListener(`error#${hash}@entries.get`, onError);

        resolve(entries);
      };

      const onError = (_: Event, errorMessage: string) => {
        this.ipcRenderer.removeListener(`respond#${hash}@entries.get`, onRespond);

        reject(new Error(errorMessage));
      };

      this.ipcRenderer.once(`respond#${hash}@entries.get`, onRespond);
      this.ipcRenderer.once(`error#${hash}@entries.get`, onError);

      this.ipcRenderer.send('entries.get', hash, text);
    });
  }

  public getPictures(text: string): Promise<string[]> {
    if (text === '') {
      return Promise.resolve([]);
    }

    return new Promise((resolve, reject) => {
      const hash = v4();

      const onRespond = (_: Event, pictureUrls: string[]) => {
        this.ipcRenderer.removeListener(`error#${hash}@pictures.get`, onError);

        resolve(pictureUrls);
      };

      const onError = (_: Event, errorMessage: string) => {
        this.ipcRenderer.removeListener(`respond#${hash}@pictures.get`, onRespond);

        reject(new Error(errorMessage));
      };

      this.ipcRenderer.once(`respond#${hash}@pictures.get`, onRespond);
      this.ipcRenderer.once(`error#${hash}@pictures.get`, onError);

      this.ipcRenderer.send('pictures.get', hash, text);
    });
  }

  public getPicture(url: URL): Promise<File> {
    return new Promise((resolve, reject) => {
      const hash = v4();

      const onRespond = (_: Event, file: Record<string, any>) => {
        this.ipcRenderer.removeListener(`error#${hash}@picture.get`, onError);

        resolve(File.fromJSON(file));
      };

      const onError = (_: Event, errorMessage: string) => {
        this.ipcRenderer.removeListener(`respond#${hash}@picture.get`, onRespond);

        reject(new Error(errorMessage));
      };

      this.ipcRenderer.once(`respond#${hash}@picture.get`, onRespond);
      this.ipcRenderer.once(`error#${hash}@picture.get`, onError);

      this.ipcRenderer.send('picture.get', hash, url.href);
    });
  }

  public synthesize(text: string): Promise<File> {
    return new Promise((resolve, reject) => {
      const hash = v4();

      const onRespond = (_: Event, base64: string) => {
        this.ipcRenderer.removeListener(`error#${hash}@synthesize`, onError);

        resolve(new File(base64, 'audio/mp3'));
      };

      const onError = (_: Event, errorMessage: string) => {
        this.ipcRenderer.removeListener(`respond#${hash}@synthesize`, onRespond);

        reject(new Error(errorMessage));
      };

      this.ipcRenderer.once(`respond#${hash}@synthesize`, onRespond);
      this.ipcRenderer.once(`error#${hash}@synthesize`, onError);

      this.ipcRenderer.send('synthesize', hash, text);
    });
  }

  public getConfiguration(): Promise<Configuration> {
    return new Promise((resolve, reject) => {
      const hash = v4();

      const onRespond = (_: Event, configuration: Configuration) => {
        this.ipcRenderer.removeListener(`error#${hash}@configuration.get`, onError);

        resolve(configuration);
      };

      const onError = (_: Event, errorMessage: string) => {
        this.ipcRenderer.removeListener(`respond#${hash}@configuration.get`, onRespond);

        reject(new Error(errorMessage));
      };

      this.ipcRenderer.once(`respond#${hash}@configuration.get`, onRespond);
      this.ipcRenderer.once(`error#${hash}@configuration.get`, onError);

      this.ipcRenderer.send('configuration.get', hash);
    });
  }

  public setConfiguration(configuration: Configuration): Promise<void> {
    return new Promise((resolve, reject) => {
      const hash = v4();

      const onRespond = (_: Event) => {
        this.ipcRenderer.removeListener(`error#${hash}@configuration.set`, onError);

        resolve();
      };

      const onError = (_: Event, errorMessage: string) => {
        this.ipcRenderer.removeListener(`respond#${hash}@configuration.set`, onRespond);

        reject(new Error(errorMessage));
      };

      this.ipcRenderer.once(`respond#${hash}@configuration.set`, onRespond);
      this.ipcRenderer.once(`error#${hash}@configuration.set`, onError);

      this.ipcRenderer.send('configuration.set', hash, configuration);
    });
  }

  public constructor(ipcRenderer: IpcRenderer) {
    this.ipcRenderer = ipcRenderer;
  }
}

export default Backend;
