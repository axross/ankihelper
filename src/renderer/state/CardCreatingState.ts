import { Container } from 'unstated';
import File from '../../common/model/File';

export type State = {
  sessionId: number;
  keyword: string;
  keywordPronunciation: File | null;
  keywordPronunciationLoading: number;
  ipa: string;
  ipaLoading: number;
  partOfSpeech: string;
  definition: string;
  example: string;
  examplePronunciation: File | null;
  examplePronunciationLoading: number;
  exampleWithBlank: string;
  imageSearchQuery: string;
  picture: File | null;
  pictureLoading: number;
  keepExample: boolean;
  cardCreating: number;
};

class CardCreatingState extends Container<State> {
  private readonly defaultState: State;
  public readonly state: State;

  public setKeyword(keyword: string) {
    this.setState({ keyword });
  }

  public setKeywordPronunciation(file: File) {
    this.setState({ keywordPronunciation: file });
  }

  public setIpa(ipa: string) {
    this.setState({ ipa });
  }

  public clearIpa() {
    this.setIpa('');
  }

  public beginLoadingIpa() {
    this.setState(({ ipaLoading }) => ({ ipaLoading: ipaLoading + 1 }));
  }

  public endLoadingIpa() {
    this.setState(({ ipaLoading }) => ({ ipaLoading: ipaLoading - 1 }));
  }

  public setPartOfSpeech(partOfSpeech: string) {
    this.setState({ partOfSpeech });
  }

  public setDefinition(definition: string) {
    this.setState({ definition });
  }

  public setExample(example: string) {
    this.setState({ example });
  }

  public setExamplePronunciation(file: File) {
    this.setState({ examplePronunciation: file });
  }

  public setExampleWithBlank(exampleWithBlank: string) {
    this.setState({ exampleWithBlank });
  }

  public setImageSearchQuery(imageSearchQuery: string) {
    this.setState({ imageSearchQuery });
  }

  public beginLoadingPicture() {
    this.setState(({ pictureLoading }) => ({ pictureLoading: pictureLoading + 1 }));
  }

  public endLoadingPicture() {
    this.setState(({ pictureLoading }) => ({ pictureLoading: pictureLoading - 1 }));
  }

  public setPicture(picture: File) {
    this.setState({ picture });
  }

  public setKeepExample(keepExample: boolean) {
    this.setState({ keepExample });
  }

  public beginCreatingCard() {
    this.setState(({ cardCreating }) => ({ cardCreating: cardCreating + 1 }));
  }

  public endCreatingCard() {
    this.setState(({ cardCreating }) => ({ cardCreating: cardCreating - 1 }));
  }

  public clearAll() {
    const sessionId =
      this.state.sessionId < Number.MAX_SAFE_INTEGER ? this.state.sessionId + 1 : this.defaultState.sessionId;

    this.setState({
      ...this.defaultState,
      sessionId,
      keepExample: this.state.keepExample,
      ...(this.state.keepExample
        ? {
            example: this.state.example,
            examplePronunciation: this.state.examplePronunciation,
            exampleWithBlank: this.state.exampleWithBlank,
            imageSearchQuery: this.state.imageSearchQuery,
            picture: this.state.picture,
          }
        : {}),
    });
  }

  public isReadyToCreate(): boolean {
    if (this.state.keyword === '') return false;
    if (this.state.keywordPronunciation === null) return false;
    if (this.state.ipa === '') return false;
    if (this.state.partOfSpeech === '') return false;
    if (this.state.definition === '') return false;
    if (this.state.example === '') return false;
    if (this.state.examplePronunciation === null) return false;
    if (this.state.exampleWithBlank === '') return false;
    if (this.state.picture === null) return false;
    if (this.state.pictureLoading >= 1) return false;
    if (this.state.cardCreating >= 1) return false;

    return true;
  }

  public constructor() {
    super();

    this.defaultState = {
      sessionId: 1,
      keyword: '',
      keywordPronunciation: null,
      keywordPronunciationLoading: 0,
      ipa: '',
      ipaLoading: 0,
      partOfSpeech: '',
      definition: '',
      example: '',
      examplePronunciation: null,
      examplePronunciationLoading: 0,
      exampleWithBlank: '',
      imageSearchQuery: '',
      picture: null,
      pictureLoading: 0,
      keepExample: false,
      cardCreating: 0,
    };
    this.state = { ...this.defaultState };
  }
}

export default CardCreatingState;
