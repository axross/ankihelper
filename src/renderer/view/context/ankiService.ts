import { Consumer as C, createContext } from 'react';
import AnkiService from '../../service/AnkiService';

const context = createContext<AnkiService>(null as any);

export const Provider = context.Provider;
export const Consumer = context.Consumer as C<AnkiService>;
