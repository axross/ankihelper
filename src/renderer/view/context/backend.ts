import { Consumer as C, createContext } from 'react';
import Backend from '../../service/Backend';

const context = createContext<Backend>(null as any);

export const Provider = context.Provider;
export const Consumer = context.Consumer as C<Backend>;
