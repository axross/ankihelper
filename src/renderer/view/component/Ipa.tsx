import { InputGroup, Intent, Toaster } from '@blueprintjs/core';
import { ChangeEvent, createElement } from 'react';
import Redebounce from 'redebounce';
import { Subscribe } from 'unstated';
import CardCreatingState from '../../state/CardCreatingState';
import { Consumer as BackendConsumer } from '../context/backend';
import Repromised from './Repromised';

type Props = {
  className?: string;
};

const Ipa = ({ className }: Props) => (
  <BackendConsumer>
    {backend => (
      <Subscribe to={[CardCreatingState]}>
        {(cardCreatingState: CardCreatingState) => (
          <Redebounce dueTime={600} value={cardCreatingState.state.keyword}>
            {keyword => (
              <Repromised
                promise={() => backend.getIpa(keyword)}
                initial=""
                then={ipa => console.log('then', ipa) || cardCreatingState.setIpa(ipa)}
                catch={err => Toaster.create().show({ intent: Intent.DANGER, message: err.message })}
                key={keyword}
              >
                {(ipa, isLoading) => (
                  <InputGroup
                    defaultValue={ipa}
                    onChange={(e: ChangeEvent<HTMLInputElement>) => cardCreatingState.setIpa(e.currentTarget.value)}
                    disabled={isLoading}
                    className={className}
                    key={keyword}
                  />
                )}
              </Repromised>
            )}
          </Redebounce>
        )}
      </Subscribe>
    )}
  </BackendConsumer>
);

export default Ipa;
