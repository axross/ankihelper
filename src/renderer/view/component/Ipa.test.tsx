import { createElement } from 'react';
import { create } from 'react-test-renderer';
import Backend from '../../service/Backend';
import CardCreatingState from '../../state/CardCreatingState';
import { Provider as BackendProvider } from '../context/backend';
import { Provider as StateProvider } from 'unstated';
import Ipa from './Ipa';

describe('<Ipa>', () => {
  describe('calls cardCreatingState.getIpa() and backend.setIpa()', () => {
    test('when it is mounted', done => {
      const backend = new Backend(Symbol('ipcRenderer') as any);
      const cardCreatingState = new CardCreatingState();
      const getIpa = jest.spyOn(backend, 'getIpa').mockImplementation((keyword: any) => Promise.resolve(keyword));
      const setIpa = jest.spyOn(cardCreatingState, 'setIpa');

      const renderer = create(
        <BackendProvider value={backend}>
          <StateProvider inject={[cardCreatingState]}>
            <Ipa />
          </StateProvider>
        </BackendProvider>
      );

      setImmediate(() => {
        expect(getIpa).toHaveBeenCalledWith('');
        expect(setIpa).toHaveBeenCalledWith('');
        expect(renderer.toJSON()).toMatchSnapshot();

        done();
      });
    });

    test('whenever cardCreatingState.state.keyword is changed', async () => {
      jest.useFakeTimers();

      const backend = new Backend(Symbol('ipcRenderer') as any);
      const cardCreatingState = new CardCreatingState();
      const getIpa = jest
        .spyOn(backend, 'getIpa')
        .mockImplementation((keyword: any) => Promise.resolve(`ipa of ${keyword}`));
      const setIpa = jest.spyOn(cardCreatingState, 'setIpa');

      const renderer = create(
        <BackendProvider value={backend}>
          <StateProvider inject={[cardCreatingState]}>
            <Ipa />
          </StateProvider>
        </BackendProvider>
      );

      await cardCreatingState.setState({ keyword: 'next keyword' });
      jest.advanceTimersByTime(600);
      jest.useRealTimers();
      await new Promise(resolve => setImmediate(resolve));
      jest.useFakeTimers();
      expect(renderer.toJSON()).toMatchSnapshot();

      await cardCreatingState.setState({ keyword: 'third something' });
      jest.advanceTimersByTime(600);
      jest.useRealTimers();
      await new Promise(resolve => setImmediate(resolve));
      jest.useFakeTimers();
      expect(renderer.toJSON()).toMatchSnapshot();

      jest.useRealTimers();

      return new Promise(resolve =>
        setImmediate(() => {
          expect(getIpa).toHaveBeenNthCalledWith(2, 'next keyword');
          expect(getIpa).toHaveBeenNthCalledWith(3, 'third something');

          expect(setIpa).toHaveBeenNthCalledWith(2, 'ipa of next keyword');
          expect(setIpa).toHaveBeenNthCalledWith(3, 'ipa of third something');

          resolve();
        })
      );
    });
  });
});
