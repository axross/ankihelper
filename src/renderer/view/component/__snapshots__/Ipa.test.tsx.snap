// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Ipa> calls cardCreatingState.getIpa() and backend.setIpa() when it is mounted 1`] = `
<div
  className="bp3-input-group"
>
  <input
    className="bp3-input"
    defaultValue=""
    disabled={false}
    onChange={[Function]}
    style={
      Object {
        "paddingRight": 10,
      }
    }
    type="text"
  />
</div>
`;

exports[`<Ipa> calls cardCreatingState.getIpa() and backend.setIpa() whenever cardCreatingState.state.keyword is changed 1`] = `
<div
  className="bp3-input-group"
>
  <input
    className="bp3-input"
    defaultValue="ipa of next keyword"
    disabled={false}
    onChange={[Function]}
    style={
      Object {
        "paddingRight": 10,
      }
    }
    type="text"
  />
</div>
`;

exports[`<Ipa> calls cardCreatingState.getIpa() and backend.setIpa() whenever cardCreatingState.state.keyword is changed 2`] = `
<div
  className="bp3-input-group"
>
  <input
    className="bp3-input"
    defaultValue="ipa of third something"
    disabled={false}
    onChange={[Function]}
    style={
      Object {
        "paddingRight": 10,
      }
    }
    type="text"
  />
</div>
`;
